### ISSUE TITLE
support filter based on magic bytes

### ISSUE THREAD

**Issue by jonassmedegaard** [Genivia/ugrep#34]
Would be nice if possible to express a filter to be used not based on extension but on magic bytes.  Maybe the elegant approach for that is to express which extension a certain set of magic bytes should internally be treated as (similar to how stdin can be assigned a virtual extension).

My use-case is scanning for copyright and license information in source code, some of which is binary data.  Concretely, some fonts - so-called Postscript Type1 fonts - are binary and without a standardized extension. I would like to be able to scan all of Ghostscript source, extracting metadata from PNG images and fonts using exiftool.

**Comment by genivia-inc** [comment]
Thanks for the feedback and the suggestion.

I like the idea of extending `--filter`. However, my initial hesitation is that it might be cumbersome to specify extensions and magic bytes together or separately for each filter choice in the list, e.g. we probably need more commas or (semi)colons to tell them apart. I just seems overly complex to achieve this featrure compared to just extension names followed by a command. By extending the `--filter` option to allow arbitrary magic byte patterns this option easily becomes error prone to small mistakes that affect its usability (and perhaps reliability).

The ability to specify magic bytes with the `--filter` option would be nice, though not essential because the same can be produced in two ways:
- use multiple searches, each for a specific combination of magic bytes and a filter. To do so, run with `-M` with a string of magic bytes and a `--filter` to filter the file matched by the magic bytes and the pattern.
- use a script specified as a `--filter` value. The script checks (e.g. with another ugrep invocation with an empty pattern to match all) the magic bytes of its file argument supplied by ugrep, then invokes the converter/output if the magic bytes match. Otherwise it produces no output.

**Comment by jonassmedegaard** [comment]
Quoting Robert van Engelen (2020-05-20 22:44:52)
> it might be cumbersome to specify extensions and magic bytes together

Sorry, I did not mean to suggest extending the --filter option itself.
I agree that would be cludgy.

> just seems overly complex to achieve this featrure compared to just extension names followed by a command.

Yes, I agree.  I was thinking a syntax like this:

ugrep -w -i 'license' --filter=png,t1:'exiftool %' --file-magic-label=t1:'%!PS-AdobeFont-1.0' project

Above example greps for "license" in either plaintext or for PNG images 
and Type1 fonts the output of exiftool.

The --file-magic-label option could be repeated for other magic strings.

**Comment by genivia-inc** [comment]
It is helpful to receive suggestions and practical use cases from power users like you.

It makes sense to look into expanding the feature set for cases that cannot be absolutely covered already, at least not by combining existing options to get what you want.

> ugrep -w -i 'license' --filter=png,t1:'exiftool %' --file-magic-label=t1:'%!PS-AdobeFont-1.0' project

I see now what you were hinting at. Sorry, I wasn't sure what feature you had in mind.

Perhaps the current options should get you close to what you want, but not exactly "in one go", i.e. one run instead of running the search engine separately to search files matching certain file extensions and another set of runs to search files matching magic bytes (and possibly searching a file twice as a consequence).

Please note that option `-O` and `-M` can be used together to search files with matching extensions and files with matching magic bytes, not their intersection. If you are using only one filter, say `exiftool` then this option combo will always work in one go:

    ugrep -w -i 'license' --filter='exiftool %' -Opng -M'%!PS-AdobeFont-1.0' project

This sends PNG and PS files to `exiftool` to search its output for the word `license`.

Better yet, you may want to search PNG files by extension or magic bytes so you won't miss finding any PNG that does not have an extension. To do so use `-tPng`, where Png is capitalized to search png extension or PNG magic bytes, whichever occurs:

    ugrep -w -i 'license' --filter='exiftool %' -tPng -M'%!PS-AdobeFont-1.0' project

However, if you have multiple filter utilities, then this approach may require a separate run each per filter utility specified, by specifying the file extensions and magic bytes of files to search for the filter. That should work, but it may not be as efficient as one go.

Hint: if you are not sure what file extensions (globs) and magic byte patterns were used in a search, use option `--stats` to display a list after the search completed.

Let me know if this helps or if I am missing something in your use case that I can help with, either with suggestions as above or by expanding ugrep if there is a powerful case to be made.

**Comment by jonassmedegaard** [comment]
Quoting Robert van Engelen (2020-05-21 16:37:47)
> Perhaps the current options should get you close to what you want, but not exactly "in one go",

I appreciate the ability to break things into smaller tasks, but my 
interest here is exactly to do the examination with only a single scan.

Both for efficiency and for ease of use.

I am exploring if possible to replace my current command-line tool 
licensecheck with ugrep and the collection of regexes developed over the 
past 19 years for licensecheck.

Maybe it turns out that ugrep will not be suitable for the license 
checking needs of Debian and Fedora and Chromium (current users of 
licensecheck as far as I am aware), but regardless I think the mechanism 
has wider use - that is the reason I propose to implement it.

I apologize that I have not been more explicit about the need for a 
single-run approach - you have given some fine examples, but for 
multi-run which is not really interesting.

Please do tell if you want to to try elaborate further.

**Comment by genivia-inc** [comment]
It is not difficult or impossible to implement `--file-magic-label` given ugrep's design. I'll have something ready to try soon.

**Comment by jonassmedegaard** [comment]
Quoting Robert van Engelen (2020-05-22 15:19:16)
> It is not difficult or impossible to implement `--file-magic-label` given ugrep's design. I'll have something ready to try soon.

great!

**Comment by genivia-inc** [comment]
I have an update ready and decided to use option `--filter-magic-label` because the new option is related only to `--filter`, not to file types.

I have one question:
- Should the filter labels be used only if the file in question has no filename extension suffix? I'm inclined to answer yes, because filename extensions should take priority over magic bytes and having both compete for potentially different filters is likely going to be confusing. But that's my opinion.

**Comment by jonassmedegaard** [comment]
Quoting Robert van Engelen (2020-05-22 16:51:00)
> I have an update ready and decided to use option `--filter-magic-label` because the new option is related only to `--filter`, not to file types.

Nice!


> - Should the filter labels be used only if the file in question has no filename extension suffix? I'm inclined to answer yes, because filename extensions should take priority over magic bytes and having both compete for potentially different filters is likely going to be confusing. But that's my opinion.

Sounds sensible by _default_ to cover only files with no extension (i.e. 
containing no dot in its filename) - but I can easily imagine a need for 
overriding said default.

How about this: Add a bang (!) as suffix to an extension to have that 
extension be enforced, disregarding any filename extensions?

Like this:

ugrep -w -i 'license' --filter=png,t1:'exiftool %' --file-magic-label=t1!:'%!PS-AdobeFont-1.0' project

...which will filter a) all files named *.png or *.t1 (regardless of 
content), and b) all files with magic of Type1 fonts (regardless of 
filename - i.e. also filtering file "My No. One Font" and (wrongly named 
- containing Type1, not TrueType) file "mysecondfont.ttf").

**Comment by genivia-inc** [comment]
Sure. But instead of `!` I prefer something like `+` because `csh` and other shells expand `!` thereby wrecking havoc on the command line. That's why for `-g` either `!` or `^` may be used to negate the glob match.

This is what I have so far that is working:
~~~
    --filter-magic-label=LABEL:MAGIC
            Associate LABEL with files whose signature "magic bytes" match the
            MAGIC regex pattern.  Only files that have no filename extension
            are labeled, unless +LABEL is specified.  When LABEL matches an
            extension specified in --filter=COMMANDS, the corresponding command
            is invoked.  This option may be repeated.
~~~

**Comment by jonassmedegaard** [comment]
Quoting Robert van Engelen (2020-05-22 18:38:35)
> Sure. But instead of `!` I prefer something like `+` because `csh` and other shells expand `!`

Your call :-)

**Comment by genivia-inc** [comment]
Implemented as a new feature in v2.1.3.
